mm2src/mm2_main/src/lp_swap/taker_swap.rs

```diff
@@ -1905,86 +1906,80 @@ impl TakerSwap {
             panic!("{}", REFUND_TEST_FAILURE_LOG);
         }
 
-        let taker_payment = self.r().taker_payment.clone().unwrap().tx_hex;
-        let locktime = self.r().data.taker_payment_lock;
-        if self.taker_coin.is_auto_refundable() {
-            return match self.taker_coin.wait_for_htlc_refund(&taker_payment, locktime).await {
-                Ok(()) => Ok((Some(TakerSwapCommand::FinalizeTakerPaymentRefund), vec![
-                    TakerSwapEvent::TakerPaymentRefunded(None),
-                ])),
-                Err(err) => Ok((Some(TakerSwapCommand::Finish), vec![
-                    TakerSwapEvent::TakerPaymentRefundFailed(
-                        ERRL!("!taker_coin.wait_for_htlc_refund: {}", err.to_string()).into(),
-                    ),
-                ])),
-            };
-        }
-
+        // Keep trying to recover funds (by refunding the taker payment or spending the maker payment) until successful or face an irrecoverable error.
         loop {
-            match self.taker_coin.can_refund_htlc(locktime).await {
-                Ok(CanRefundHtlc::CanRefundNow) => break,
-                Ok(CanRefundHtlc::HaveToWait(to_sleep)) => Timer::sleep(to_sleep as f64).await,
-                Err(e) => {
-                    error!("Error {} on can_refund_htlc, retrying in 30 seconds", e);
-                    Timer::sleep(30.).await;
+            match self.recover_funds().await {
+                // We recovered the swap successfully.
+                Ok(recovered_swap) => {
+                    let tx_ident = TransactionIdentifier {
+                        tx_hex: recovered_swap.transaction.tx_hex().into(),
+                        tx_hash: recovered_swap.transaction.tx_hash_as_bytes(),
+                    };
+                    return match recovered_swap.action {
+                        // We recovered the swap by refunding the taker payment.
+                        RecoveredSwapAction::RefundedMyPayment => {
+                            info!("Taker payment refund tx {:02x}", tx_ident.tx_hash);
+                            Ok((Some(TakerSwapCommand::FinalizeTakerPaymentRefund), vec![
+                                TakerSwapEvent::TakerPaymentRefunded(Some(tx_ident)),
+                            ]))
+                        },
+                        // We recovered the swap by proceeding forward and spending the maker payment. The swap wasn't actually a failure.
+                        // Roll back to confirming the maker payment spend.
+                        RecoveredSwapAction::SpentOtherPayment => {
+                            info!("Refund canceled. Maker payment spend tx {:02x}", tx_ident.tx_hash);
+                            // TODO: We prepared for refund but didn't finalize refund. This must be breaking something for lightning.
```

Comment:
Will we ever hit this code in lightning? I guess it can happen for taker swap code but it's not a big problem anyways since the finalize refund is just to allow the maker get his payment back instantly instead of waiting for the timelock to expire, and if we spend it, it means it can't be failed backwards. For maker swap code it's a different case, maker fails the htlc backwards before this step, so we will never hit this code.
Usually such things in code is a cue for a needed refactor but not in this PR of course.